File: luminus/src/cards_project/content/article_handler.clj
(ns cards_project.content.article-handler
  (:require [compojure.core :refer [defroutes GET POST PUT PATCH DELETE]]
            [ring.util.response :as resp]
            [next.jdbc :as jdbc]
            [next.jdbc.result-set :as rs]
            [cards_project.content.article-queries :as queries]
            [cards_project.content.article-service :as svc]
            [cards_project.db :refer [db-spec]]))

(defn- article-kw-params [params]
  (into {} (map (fn [[k v]] [(keyword (clojure.string/replace (name k) "-" "_")) v]) params)))

(defn- ->num [v] (when (some? v) (if (string? v) (Double/parseDouble v) (double v))))

(defn- validate-article-rules! [m]
  (let [errors (atom [])]
    (when-not (let [v (get m :view_count)] (or (nil? v) (>= (->num v) 0)))
      (swap! errors conj "Article view count must not be negative"))
    (when-not (let [v (get m :likes_count)] (or (nil? v) (>= (->num v) 0)))
      (swap! errors conj "Article likes count must not be negative"))
    (when (seq @errors)
      (throw (ex-info "Validation failed" {:errors @errors :status 422})))))

(defn- validate-article-implies! [m]
  (let [errors (atom [])]
    (when (and (= (get m :status) "Published") (not (some? (get m :published_at))))
      (swap! errors conj "Published article must have a published_at timestamp"))
    (when (seq @errors)
      (throw (ex-info "Validation failed" {:errors @errors :status 422})))))

(defn- insert-article! [params]
  (let [kw-params (into {} (map (fn [[k v]] [(keyword (clojure.string/replace (name k) "-" "_")) v]) params))
        allowed  #{:title :slug :body :excerpt :cover_image_url :status :article_type :language :view_count :likes_count :is_featured :published_at :author_id :featured_deck_id}
        pairs    (filter (fn [[k _]] (allowed k)) kw-params)
        cols     (map #(name (first %)) pairs)
        vals     (map second pairs)
        sql      (str "INSERT INTO articles ("
                     (clojure.string/join ", " cols)
                     ") VALUES ("
                     (clojure.string/join ", " (repeat (count cols) "?"))
                     ")")]
    (with-open [conn (jdbc/get-connection db-spec)]
      (jdbc/execute-one! conn (into [sql] vals))
      (-> (jdbc/execute-one! conn ["SELECT last_insert_rowid() AS id"]
                             {:builder-fn rs/as-unqualified-lower-maps})
          :id))))

(defn- update-article! [id params]
  (let [kw-params (into {} (map (fn [[k v]] [(keyword (clojure.string/replace (name k) "-" "_")) v]) params))
        allowed  #{:title :slug :body :excerpt :cover_image_url :status :article_type :language :view_count :likes_count :is_featured :published_at :author_id :featured_deck_id}
        pairs    (filter (fn [[k _]] (allowed k)) kw-params)
        cols     (map #(name (first %)) pairs)
        vals     (map second pairs)
        sql      (str "UPDATE articles SET "
                     (clojure.string/join ", " (map #(str % " = ?") cols))
                     " WHERE id = ?")]
    (jdbc/execute-one! db-spec (into [sql] (conj (vec vals) id)))))

(defn- apply-projection-article [record]
  (when record
    (let [m (let [m (let [m record] (-> m (dissoc :created_at) (assoc :created_at (get m :created_at))))] (-> m (dissoc :updated_at) (assoc :updated_at (get m :updated_at))))] (-> m (dissoc :published_at) (assoc :published_at (get m :published_at))))))

(defroutes articles-routes

  (GET "/api/articles" {params :query-params}
    (let [q (or (get params "q") "")]
      (resp/response (map apply-projection-article (filter #(or (empty? q) (or (clojure.string/includes? (str (get % :title "")) q) (clojure.string/includes? (str (get % :excerpt "")) q))) (queries/get-all-article db-spec))))))

  (POST "/api/articles" {params :body}
    (try
      (let [kw (article-kw-params params)]
        (validate-article-rules! kw)
        (validate-article-implies! kw)
        (let [new-id (insert-article! params)
              record  (or (queries/get-article-by-id db-spec {:id new-id}) {:id new-id})]
          (-> (resp/response (apply-projection-article record)) (resp/status 201))))
      (catch clojure.lang.ExceptionInfo e
        (-> (resp/response {:errors (:errors (ex-data e))}) (resp/status 422)))
      (catch Exception e
        (-> (resp/response {:error (.getMessage e)}) (resp/status 500)))))

  (GET "/api/articles/:id" [id]
    (if-let [record (queries/get-article-by-id db-spec {:id (Integer/parseInt id)})]
      (resp/response (apply-projection-article record))
      (-> (resp/response {:error "Not found"}) (resp/status 404))))

  (PUT "/api/articles/:id" [id :as {params :body :as req}]
    (try
      (let [kw (article-kw-params params)]
        (validate-article-rules! kw)
        (validate-article-implies! kw)
        (let [int-id (Integer/parseInt id)]
          (update-article! int-id params)
          (if-let [record (queries/get-article-by-id db-spec {:id int-id})]
            (resp/response (apply-projection-article record))
            (-> (resp/response {:error "Not found"}) (resp/status 404)))))
      (catch clojure.lang.ExceptionInfo e
        (-> (resp/response {:errors (:errors (ex-data e))}) (resp/status 422)))
      (catch Exception e
        (-> (resp/response {:error (.getMessage e)}) (resp/status 500)))))

  (PATCH "/api/articles/:id" [id :as {params :body :as req}]
    (try
      (let [kw (article-kw-params params)]
        (validate-article-rules! kw)
        (validate-article-implies! kw)
        (let [int-id (Integer/parseInt id)]
          (update-article! int-id params)
          (if-let [record (queries/get-article-by-id db-spec {:id int-id})]
            (resp/response (apply-projection-article record))
            (-> (resp/response {:error "Not found"}) (resp/status 404)))))
      (catch clojure.lang.ExceptionInfo e
        (-> (resp/response {:errors (:errors (ex-data e))}) (resp/status 422)))
      (catch Exception e
        (-> (resp/response {:error (.getMessage e)}) (resp/status 500)))))


  (POST "/api/articles/:id/publish" [id]
    (svc/publish! (Integer/parseInt id))
    (-> (resp/response nil) (resp/status 204)))

  (POST "/api/articles/:id/archive" [id]
    (svc/archive! (Integer/parseInt id))
    (-> (resp/response nil) (resp/status 204)))

  (POST "/api/articles/:id/view" [id]
    (svc/increment-view! (Integer/parseInt id))
    (-> (resp/response nil) (resp/status 204)))

  (POST "/api/articles/:id/like" [id]
    (svc/like! (Integer/parseInt id))
    (-> (resp/response nil) (resp/status 204)))

  (DELETE "/api/articles/:id/like" [id]
    (svc/unlike! (Integer/parseInt id))
    (-> (resp/response nil) (resp/status 204)))

  (GET "/api/articles/:id/reading-time" [id]
    (let [result (svc/reading-time-minutes! (Integer/parseInt id))]
      (resp/response {:result result})))

  (PATCH "/api/articles/:id/transitions/draft-to-published" [id :as request]
    (let [user-role (get-in request [:headers "x-user-role"])]
      (if (not (contains? #{"Editor" "Admin"} user-role))
        (-> (resp/response {:error "Insufficient role for transition Draft -> Published"}) (resp/status 403))
        (try
          (let [record (svc/transition-draft-to-published! (Integer/parseInt id))]
            (resp/response record))
          (catch clojure.lang.ExceptionInfo e
            (let [status (get (ex-data e) :status 500)]
              (-> (resp/response {:error (.getMessage e)}) (resp/status status))))
          (catch Exception e
            (-> (resp/response {:error (.getMessage e)}) (resp/status 500)))))))

  (PATCH "/api/articles/:id/transitions/published-to-archived" [id :as request]
    (let [user-role (get-in request [:headers "x-user-role"])]
      (if (not (contains? #{"Editor" "Admin"} user-role))
        (-> (resp/response {:error "Insufficient role for transition Published -> Archived"}) (resp/status 403))
        (try
          (let [record (svc/transition-published-to-archived! (Integer/parseInt id))]
            (resp/response record))
          (catch clojure.lang.ExceptionInfo e
            (let [status (get (ex-data e) :status 500)]
              (-> (resp/response {:error (.getMessage e)}) (resp/status status))))
          (catch Exception e
            (-> (resp/response {:error (.getMessage e)}) (resp/status 500)))))))

  (PATCH "/api/articles/:id/transitions/archived-to-draft" [id :as request]
    (let [user-role (get-in request [:headers "x-user-role"])]
      (if (not (contains? #{"Admin"} user-role))
        (-> (resp/response {:error "Insufficient role for transition Archived -> Draft"}) (resp/status 403))
        (try
          (let [record (svc/transition-archived-to-draft! (Integer/parseInt id))]
            (resp/response record))
          (catch clojure.lang.ExceptionInfo e
            (let [status (get (ex-data e) :status 500)]
              (-> (resp/response {:error (.getMessage e)}) (resp/status status))))
          (catch Exception e
            (-> (resp/response {:error (.getMessage e)}) (resp/status 500)))))))

  (PATCH "/api/articles/:id/transitions/published-to-draft" [id]
    (try
      (let [record (svc/transition-published-to-draft! (Integer/parseInt id))]
        (resp/response record))
      (catch clojure.lang.ExceptionInfo e
        (let [status (get (ex-data e) :status 500)]
          (-> (resp/response {:error (.getMessage e)}) (resp/status status))))
      (catch Exception e
        (-> (resp/response {:error (.getMessage e)}) (resp/status 500)))))
)

File: dream/lib/cards_project/content/article_handler.ml
(* Dream handlers for Article *)
open Lwt.Syntax

let apply_projection_article (j : Yojson.Safe.t) : Yojson.Safe.t =
  match j with
  | `Assoc fields ->
    let fields = List.filter (fun (k, _) -> not (List.mem k [])) fields in
    let fields = List.map (fun (k, v) -> if k = "created_at" then ("created_at", v) else (k, v)) fields in
    let fields = List.map (fun (k, v) -> if k = "updated_at" then ("updated_at", v) else (k, v)) fields in
    let fields = List.map (fun (k, v) -> if k = "published_at" then ("published_at", v) else (k, v)) fields in
    `Assoc fields
  | other -> other

(* JSON field helpers *)
let json_string_opt j key =
  match Yojson.Safe.Util.member key j with
  | `String s -> Some s
  | `Null | `Assoc [] -> None
  | _ -> None

let json_float_opt j key =
  match Yojson.Safe.Util.member key j with
  | `Float f -> Some f
  | `Int i -> Some (float_of_int i)
  | `String s -> (try Some (float_of_string s) with _ -> None)
  | _ -> None

let json_present j key =
  match Yojson.Safe.Util.member key j with
  | `Null | `Assoc [] -> false
  | _ -> true

let json_bool_opt j key =
  match Yojson.Safe.Util.member key j with
  | `Bool b -> Some b
  | _ -> None

let validate_article (j : Yojson.Safe.t) : (unit, string list) result =
  let errors = ref [] in
  if not ((match (json_float_opt j "view_count") with Some v -> v >= 0. | None -> true)) then errors := "Article view count must not be negative" :: !errors;
  if not ((match (json_float_opt j "likes_count") with Some v -> v >= 0. | None -> true)) then errors := "Article likes count must not be negative" :: !errors;
  if not ((not ((json_string_opt j "status") = Some "Published") || ((json_present j "published_at")))) then errors := "Published article must have a published_at timestamp" :: !errors;
  if !errors = [] then Ok () else Error (List.rev !errors)

let extract_insert_params (j : Yojson.Safe.t) =
  let open Yojson.Safe.Util in
  let title = match member "title" j with `String s -> s | _ -> "" in
  let slug = match member "slug" j with `String s -> s | _ -> "" in
  let body = match member "body" j with `String s -> s | _ -> "" in
  let excerpt = match member "excerpt" j with `String s -> Some s | _ -> None in
  let cover_image_url = match member "cover_image_url" j with `String s -> Some s | _ -> None in
  let status = match member "status" j with `String s -> s | _ -> "" in
  let article_type = match member "article_type" j with `String s -> s | _ -> "" in
  let language = match member "language" j with `String s -> s | _ -> "" in
  let view_count = match member "view_count" j with `Int i -> i | _ -> 0 in
  let likes_count = match member "likes_count" j with `Int i -> i | _ -> 0 in
  let is_featured = match member "is_featured" j with `Bool b -> b | _ -> false in
  let published_at = match member "published_at" j with `String s -> Some s | _ -> None in
  let author_id = match member "author_id" j with `Int i -> i | _ -> 0 in
  let featured_deck_id = match member "featured_deck_id" j with `Int i -> Some i | _ -> None in
  ((title, slug, body, excerpt), (cover_image_url, status, article_type, language), (view_count, likes_count, is_featured, published_at), (author_id, featured_deck_id))

let handler_article (db : (module Caqti_lwt.CONNECTION)) req =
  let respond_json status body =
    Dream.respond ~status:(Dream.int_to_status status)
      ~headers:[("Content-Type", "application/json")]
      (Yojson.Safe.to_string body)
  in
  let module Db = (val db) in
  let target_path = match String.index_opt (Dream.target req) '?' with
    | Some i -> String.sub (Dream.target req) 0 i
    | None -> Dream.target req
  in
  let path_segments = target_path |> String.split_on_char '/' |> List.filter (fun s -> s <> "") in
  match Dream.method_ req, path_segments with

  (* GET /api/articles - list with optional ?q= search *)
  | `GET, ["api"; "articles"] ->
    let q = Dream.query req "q" |> Option.value ~default:"" in
    let* rows = Db.collect_list Article_model.get_all_q () in
    (match rows with
     | Error e -> respond_json 500 (`String (Caqti_error.show e))
     | Ok items ->
       let filtered = if q = "" then items
         else List.filter (fun r ->
           let s = (Article_model.row_to_t r).title in String.length s > 0 && (try ignore (Str.search_forward (Str.regexp_string q) s 0); true with Not_found -> false)
           || (match (Article_model.row_to_t r).excerpt with Some s -> String.length s > 0 && (try ignore (Str.search_forward (Str.regexp_string q) s 0); true with Not_found -> false) | None -> false)) items in
       let json = `List (List.map (fun r ->
         let j = Article_model.to_yojson (Article_model.row_to_t r) in
         apply_projection_article j) filtered) in
       respond_json 200 json)

  (* POST /api/articles - create *)
  | `POST, ["api"; "articles"] ->
    let* body = Dream.body req in
    (try
       let j = Yojson.Safe.from_string body in
       (match validate_article j with
        | Error errs -> respond_json 422 (`Assoc [("errors", `List (List.map (fun e -> `String e) errs))])
        | Ok () ->
       let params = extract_insert_params j in
       let* ins = Db.exec Article_model.insert_q params in
       (match ins with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok () ->
          let* last_id = Db.find Article_model.last_id_q () in
          (match last_id with
           | Error e -> respond_json 500 (`String (Caqti_error.show e))
           | Ok new_id ->
             let* row = Db.find_opt Article_model.get_by_id_q new_id in
             (match row with
              | Error e -> respond_json 500 (`String (Caqti_error.show e))
              | Ok (Some r) ->
                let j = Article_model.to_yojson (Article_model.row_to_t r) in
                respond_json 201 (apply_projection_article j)
              | Ok None -> respond_json 404 (`String "Not found")))))
     with _ -> respond_json 400 (`String "Invalid JSON"))

  (* GET /api/articles/:id - get one *)
  | `GET, ["api"; "articles"; id_str] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* row = Db.find_opt Article_model.get_by_id_q id in
       (match row with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok None -> respond_json 404 (`String "Not found")
        | Ok (Some r) ->
          let j = Article_model.to_yojson (Article_model.row_to_t r) in
          respond_json 200 (apply_projection_article j)))

  (* PUT /api/articles/:id - full update *)
  | `PUT, ["api"; "articles"; id_str] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* body = Dream.body req in
       (try
          let j = Yojson.Safe.from_string body in
          (match validate_article j with
           | Error errs -> respond_json 422 (`Assoc [("errors", `List (List.map (fun e -> `String e) errs))])
           | Ok () ->
          let params = extract_insert_params j in
          let ((title, slug, body, excerpt), (cover_image_url, status, article_type, language), (view_count, likes_count, is_featured, published_at), (author_id, featured_deck_id)) = params in
          let upd_params = ((title, slug, body, excerpt), (cover_image_url, status, article_type, language), (view_count, likes_count, is_featured, published_at), (author_id, featured_deck_id, id)) in
          let* upd = Db.exec Article_model.update_q upd_params in
          (match upd with
           | Error e -> respond_json 500 (`String (Caqti_error.show e))
           | Ok () ->
             let* row = Db.find_opt Article_model.get_by_id_q id in
             (match row with
              | Error e -> respond_json 500 (`String (Caqti_error.show e))
              | Ok None -> respond_json 404 (`String "Not found")
              | Ok (Some r) ->
                let j = Article_model.to_yojson (Article_model.row_to_t r) in
                respond_json 200 (apply_projection_article j))))
       with _ -> respond_json 400 (`String "Invalid JSON")))

  (* ── Lifecycle transitions ── *)
  | `PATCH, ["api"; "articles"; id_str; "transitions"; "draft-to-published"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* row = Db.find_opt Article_model.get_by_id_q id in
       (match row with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok None -> respond_json 404 (`String "Not found")
        | Ok (Some r) ->
          let rec_ = Article_model.row_to_t r in
          if rec_.status <> "Draft" then
            respond_json 409 (`String "Invalid transition")
          else
            (* TODO: update status to Published via DB *)
            respond_json 200 (`Assoc [("status", `String "Published")])))

  | `PATCH, ["api"; "articles"; id_str; "transitions"; "published-to-archived"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* row = Db.find_opt Article_model.get_by_id_q id in
       (match row with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok None -> respond_json 404 (`String "Not found")
        | Ok (Some r) ->
          let rec_ = Article_model.row_to_t r in
          if rec_.status <> "Published" then
            respond_json 409 (`String "Invalid transition")
          else
            (* TODO: update status to Archived via DB *)
            respond_json 200 (`Assoc [("status", `String "Archived")])))

  | `PATCH, ["api"; "articles"; id_str; "transitions"; "archived-to-draft"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* row = Db.find_opt Article_model.get_by_id_q id in
       (match row with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok None -> respond_json 404 (`String "Not found")
        | Ok (Some r) ->
          let rec_ = Article_model.row_to_t r in
          if rec_.status <> "Archived" then
            respond_json 409 (`String "Invalid transition")
          else
            (* TODO: update status to Draft via DB *)
            respond_json 200 (`Assoc [("status", `String "Draft")])))

  | `PATCH, ["api"; "articles"; id_str; "transitions"; "published-to-draft"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some id ->
       let* row = Db.find_opt Article_model.get_by_id_q id in
       (match row with
        | Error e -> respond_json 500 (`String (Caqti_error.show e))
        | Ok None -> respond_json 404 (`String "Not found")
        | Ok (Some _) ->
          respond_json 409 (`String "Transition Published -> Draft not allowed")))

  (* POST /api/articles/{id}/publish - behavior publish *)
  | `POST, ["api"; "articles"; id_str; "_id/publish"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior publish *)
       respond_json 204 (`Null))

  (* POST /api/articles/{id}/archive - behavior archive *)
  | `POST, ["api"; "articles"; id_str; "_id/archive"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior archive *)
       respond_json 204 (`Null))

  (* POST /api/articles/{id}/view - behavior increment_view *)
  | `POST, ["api"; "articles"; id_str; "_id/view"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior increment_view *)
       respond_json 204 (`Null))

  (* POST /api/articles/{id}/like - behavior like *)
  | `POST, ["api"; "articles"; id_str; "_id/like"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior like *)
       respond_json 204 (`Null))

  (* DELETE /api/articles/{id}/like - behavior unlike *)
  | `GET, ["api"; "articles"; id_str; "_id/like"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior unlike *)
       respond_json 204 (`Null))

  (* GET /api/articles/{id}/reading-time - behavior reading_time_minutes *)
  | `GET, ["api"; "articles"; id_str; "_id/reading-time"] ->
    (match int_of_string_opt id_str with
     | None -> respond_json 400 (`String "Invalid id")
     | Some _id ->
       (* TODO: implement behavior reading_time_minutes *)
       respond_json 204 (`Null))

  | _ -> respond_json 404 (`String "Not found")
